lamindb/_curate.py

```diff
@@ -1058,6 +1062,490 @@ def save_artifact(
 
 
+def _maybe_validation_error(nonval_keys: list[str], name: str):
+    if (n := len(nonval_keys)) > 0:
+        s = "s" if n > 1 else ""
+        are = "are" if n > 1 else "is"
+        raise ValidationError(
+            f"the following {n} key{s} passed to {name} {are} not allowed: {colors.yellow(_print_values(nonval_keys))}"
+        )
+
+
+class SOMACurator(BaseCurator):
+    """Curation flow for ``tiledbsoma``.
+
+    See also :class:`~lamindb.Curator`.
+
+    Args:
+        experiment_uri: The uri of a `tiledbsoma.Experiment`.
+        var_index: The registry fields for mapping the ``.var`` indices for measurements.
+            Should be in the form `{"measurement name": ("var column", field)}`.
+            These keys should be used in the flattened form ('{measurement name}__{column name in .var}')
+            in `.standardize` or `.add_new_from`, see the output of `.var_index`.
+        categoricals: A dictionary mapping ``.obs`` columns to a registry field.
+        obs_columns: The registry field for mapping the ``.obs`` columns.
+        using_key: A reference LaminDB instance.
+        organism: The organism name.
+        sources: A dictionary mapping ``.obs`` columns to Source records.
+        exclude: A dictionary mapping column names to values to exclude.
+
+    Examples:
+        >>> import bionty as bt
+        >>> curate = ln.Curator.from_tiledbsoma(
+        ...     "store.tiledbsoma",
+        ...     var_index={"RNA": ("var_id", bt.Gene.symbol)},
+        ...     categoricals={
+        ...         "cell_type_ontology_id": bt.CellType.ontology_id,
+        ...         "donor_id": ln.ULabel.name
+        ...     },
+        ...     organism="human",
+        ... )
+    """
+
+    def __init__(
+        self,
+        experiment_uri: UPathStr,
+        var_index: dict[str, tuple[str, FieldAttr]],
+        categoricals: dict[str, FieldAttr] | None = None,
+        obs_columns: FieldAttr = Feature.name,
+        using_key: str | None = None,
+        organism: str | None = None,
+        sources: dict[str, Record] | None = None,
+        exclude: dict[str, str | list[str]] | None = None,
+    ):
+        self._obs_fields = categoricals or {}
+        self._var_fields = var_index
+        self._columns_field = obs_columns
+        self._experiment_uri = UPath(experiment_uri)
+        self._organism = organism
+        self._using_key = using_key
+        self._sources = sources or {}
+        self._exclude = exclude or {}
+
+        self._validated: bool | None = False
+        self._non_validated_values: dict[str, list] | None = None
+        self._validated_values: dict[str, list] = {}
+        # filled by _check_save_keys
+        self._n_obs: int | None = None
+        self._valid_obs_keys: list[str] | None = None
+        self._valid_var_keys: list[str] | None = None
+        self._var_fields_flat: dict[str, FieldAttr] | None = None
+        self._check_save_keys()
+
+    def _check_save_keys(self):
```

Comment:
What does this do? A high-level comment?

Reply:
Added the comment.